kinto/core/storage/postgresql/__init__.py

```diff
@@ -157,48 +118,68 @@ def _check_database_encoding(self):
+         WHERE table_name = 'records';
+        """
+        schema_version_metadata_query = """
         SELECT value AS version
           FROM metadata
          WHERE name = 'storage_schema_version'
          ORDER BY LPAD(value, 3, '0') DESC;
         """
         with self.client.connect() as conn:
-            result = conn.execute(query)
+            result = conn.execute(records_table_query)
+            records_table_exists = result.rowcount > 0
+
+            if not records_table_exists:
+                return
+
+            result = conn.execute(schema_version_metadata_query)
             if result.rowcount > 0:
                 return int(result.fetchone()['version'])
-            else:
-                # Guess current version.
-                query = 'SELECT COUNT(*) FROM metadata;'
-                result = conn.execute(query)
-                was_flushed = int(result.fetchone()[0]) == 0
-                if was_flushed:
-                    error_msg = 'Missing schema history: consider version {}.'
-                    logger.warning(error_msg.format(self.schema_version))
-                    return self.schema_version
-
-                # In the first versions of Cliquet, there was no migration.
+
+            # No storage_schema_version row.
+            # Perhaps it got flush()ed by a pre-8.1.2 Kinto (which
+            # would wipe the metadata table).
+            # Alternately, maybe we are working from a very early
+            # Cliquet version which never had a migration.
+            # Check for a created_at row. If this is gone, it's
+            # probably been flushed at some point.
+            query = "SELECT COUNT(*) FROM metadata WHERE name = 'created_at';"
+            result = conn.execute(query)
+            was_flushed = int(result.fetchone()[0]) == 0
+            if not was_flushed:
+                error_msg = 'No schema history; assuming migration from Cliquet (version 1).'
+                logger.warning(error_msg)
                 return 1
 
+            # We have no idea what the schema is here. Migration
+            # is completely broken.
+            # Log an obsequious error message to the user and try
+            # to recover by assuming the last version where we had
+            # this bug.
+            logger.warning(UNKNOWN_SCHEMA_VERSION_MESSAGE)
+
+            # This is the last schema version where flushing the
+            # server would delete the schema version.
+            MAX_FLUSHABLE_SCHEMA_VERSION = 20
+            return MAX_FLUSHABLE_SCHEMA_VERSION
+
     def flush(self, auth=None):
-        """Delete records from tables without destroying schema. Mainly used
-        in tests suites.
+        """Delete records from tables without destroying schema.
+
+        This is used in test suites as well as in the flush plugin.
         """
         query = """
         DELETE FROM records;
         DELETE FROM timestamps;
-        DELETE FROM metadata;
```

Comment:
So this fixes #1460

Reply:
Yes (as described in the PR description as well as the commit comments).